File: spring-cloud-openfeign/feign-provider/feign-provider-api/src/main/java/com/livk/provider/api/OpenFeignConfig.java
package com.livk.provider.api;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.livk.auto.service.annotation.SpringAutoService;
import com.livk.provider.api.feign.factory.UserRemoteServiceFallbackFactory;
import feign.Feign;
import feign.Logger;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.cache.CacheManager;
import org.springframework.cloud.openfeign.FeignAutoConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.data.redis.cache.RedisCacheConfiguration;
import org.springframework.data.redis.cache.RedisCacheManager;
import org.springframework.data.redis.cache.RedisCacheWriter;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.serializer.Jackson2JsonRedisSerializer;
import org.springframework.data.redis.serializer.RedisSerializationContext;
import org.springframework.data.redis.serializer.StringRedisSerializer;

/**
 * <p>
 * OpenFeignConfig
 * </p>
 *
 * @author livk
 * @date 2021/12/7
 */
@SpringAutoService
@AutoConfiguration(before = FeignAutoConfiguration.class)
@ConditionalOnClass(Feign.class)
public class OpenFeignConfig {

    @Bean
    Logger.Level feignLoggerLevel() {
        System.setProperty("logging.level.com.livk.provider.api.feign", "debug");
        return Logger.Level.FULL;
    }

    @Bean
    public CacheManager cacheManager(RedisConnectionFactory redisConnectionFactory) {
        RedisCacheConfiguration redisCacheConfiguration = RedisCacheConfiguration.defaultCacheConfig();
        ObjectMapper mapper = new ObjectMapper();
        mapper.setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.ANY);
        Jackson2JsonRedisSerializer<Object> serializer = new Jackson2JsonRedisSerializer<>(mapper, Object.class);
        redisCacheConfiguration = redisCacheConfiguration.disableCachingNullValues()
                .serializeKeysWith(
                        RedisSerializationContext.SerializationPair.fromSerializer(new StringRedisSerializer()))
                .serializeValuesWith(RedisSerializationContext.SerializationPair.fromSerializer(serializer));
        return RedisCacheManager.builder(RedisCacheWriter.nonLockingRedisCacheWriter(redisConnectionFactory))
                .cacheDefaults(redisCacheConfiguration).build();
    }

    @Bean
    public UserRemoteServiceFallbackFactory userRemoteServiceFallbackFactory() {
        return new UserRemoteServiceFallbackFactory();
    }
}

File: spring-cloud-openfeign/feign-provider/feign-provider-api/src/main/kotlin/com/livk/provider/api/OpenFeignConfig.kt
package com.livk.provider.api

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.PropertyAccessor
import com.livk.auto.service.annotation.SpringAutoService
import com.livk.provider.api.feign.factory.UserRemoteServiceFallbackFactory
import feign.Feign
import feign.Logger
import org.springframework.boot.autoconfigure.AutoConfiguration
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.cache.CacheManager
import org.springframework.cloud.openfeign.FeignAutoConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.data.redis.cache.RedisCacheConfiguration
import org.springframework.data.redis.cache.RedisCacheManager
import org.springframework.data.redis.cache.RedisCacheWriter
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.serializer.JacksonJsonRedisSerializer
import org.springframework.data.redis.serializer.RedisSerializationContext
import org.springframework.data.redis.serializer.StringRedisSerializer
import tools.jackson.databind.json.JsonMapper

/**
 * @author livk
 */
@SpringAutoService
@AutoConfiguration(before = [FeignAutoConfiguration::class])
@ConditionalOnClass(Feign::class)
open class OpenFeignConfig {
    @Bean
    open fun feignLoggerLevel(): Logger.Level {
        System.setProperty("logging.level.com.livk.provider.api.feign", "debug")
        return Logger.Level.FULL
    }

    @Bean
    open fun cacheManager(redisConnectionFactory: RedisConnectionFactory): CacheManager {
        var redisCacheConfiguration = RedisCacheConfiguration.defaultCacheConfig()
        val mapper = JsonMapper.builder().changeDefaultVisibility {
            it.withVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.ANY)
        }.build()
        val serializer = JacksonJsonRedisSerializer(mapper, Any::class.java)
        redisCacheConfiguration = redisCacheConfiguration.disableCachingNullValues()
            .serializeKeysWith(
                RedisSerializationContext.SerializationPair.fromSerializer(StringRedisSerializer())
            )
            .serializeValuesWith(RedisSerializationContext.SerializationPair.fromSerializer(serializer))
        return RedisCacheManager.builder(RedisCacheWriter.nonLockingRedisCacheWriter(redisConnectionFactory))
            .cacheDefaults(redisCacheConfiguration).build()
    }

    @Bean
    open fun userRemoteServiceFallbackFactory(): UserRemoteServiceFallbackFactory {
        return UserRemoteServiceFallbackFactory()
    }
}
